File: Reqnroll/Chapter21/21.1-TestEnvironmentInitializationPattern/After/WIMP.Specs/Support/Hooks.cs
using Reqnroll;

using WIMP.App.Data;
using WIMP.Specs.Drivers;

namespace WIMP.Specs.Support;

[Binding]
public class Hooks(AppHostingContext appHostingContext, IDatabaseDriver databaseDriver, IDataRepository dataRepository, MenuBackdoorDriver menuBackdoorDriver)
{
    [BeforeTestRun]
    public static void InitializeDatabase()
    {
        var testConfigurationProvider = new TestConfigurationProvider();
        if (!testConfigurationProvider.Database.UseStub)
        {
            DatabaseDriver.UpgradeSchemaIfNeeded(testConfigurationProvider.Database.ConnectionString);
        }
    }

    [BeforeScenario(Order = 0)]
    public async Task ResetDatabase()
    {
        await databaseDriver.EmptyDatabase();
        SeedMenuData();
    }

    private void SeedMenuData()
    {
        Console.WriteLine("seeding menu data");
        menuBackdoorDriver.SetMenuItems(
        [
            new MenuItemData { Name = "Margherita", Price = 7.99m, Calories = 900, Vegetarian = true },
            new MenuItemData { Name = "Pepperoni", Price = 9.99m, Calories = 1200 },
            new MenuItemData { Name = "Capricciosa", Price = 8.99m, Calories = 1100 }
        ]);
    }

    [BeforeScenario(Order = 1)]
    public void CreateAppHost()
    {
        appHostingContext.AppHost = new WimpAppHost(dataRepository);
    }

    [AfterScenario]
    public void DisposeAppHost()
    {
        appHostingContext.AppHost?.Dispose();
    }
}

File: CucumberJVM/Chapter21/21.2-ResourcePoolingPattern/After/src/test/java/com/wimp/app/specs/support/Hooks.java
package com.wimp.app.specs.support;

import com.wimp.app.specs.drivers.DatabaseDriver;
import com.wimp.app.specs.drivers.MenuBackdoorDriver;
import io.cucumber.java.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.math.BigDecimal;
import java.util.List;

public class Hooks {
    private static final Logger log = LoggerFactory.getLogger(Hooks.class);

    private final DatabaseDriver databaseDriver;
    private final MenuBackdoorDriver menuBackdoorDriver;
    private final TestConfigurationProvider testConfigurationProvider;

    public Hooks(DatabaseDriver databaseDriver, MenuBackdoorDriver menuBackdoorDriver, TestConfigurationProvider testConfigurationProvider) {
        this.databaseDriver = databaseDriver;
        this.menuBackdoorDriver = menuBackdoorDriver;
        this.testConfigurationProvider = testConfigurationProvider;
    }

    @Before(order = 0)
    public void resetDatabase() {
        log.info("Test database: use-stub={}, use-pool={}", testConfigurationProvider.testDatabase().useStub(), testConfigurationProvider.testDatabase().usePool());
        databaseDriver.emptyDatabase();
        seedMenuData();
    }

    private void seedMenuData() {
        log.info("Seeding menu data");
        menuBackdoorDriver.setMenuItems(List.of(
            new MenuItemData("Margherita", new BigDecimal("7.99"), 900, true),
            new MenuItemData("Pepperoni", new BigDecimal("9.99"), 1200, false),
            new MenuItemData("Capricciosa", new BigDecimal("8.99"), 1100, false)));
    }
}
